doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,447 @@
+of the firmware, as well as a way to persist these values. Runtime
+configurations are deployment-specific and can be changed on a per node basis.
+Appropriate management tools could also enable the configuration of node
+groups.
+
+Examples of runtime configurations are:
+- Transmission duty cycles
+- Sensor thresholds
+- Security credentials
+- System state variables
+
+These parameters might have constraints, like a specific order to be applied
+(due to interdependencies) or value boundaries.
+
+The main advantages of having such a system are:
+- Easy to apply per-node configuration during deployment
+- No need to implement a special mechanism for per-node configurations during
+  firmware updates (only in the case of migration)
+- Common interface for modules to expose their runtime configuration parameters
+  and handle them
+- Transparent interface for storing configuration parameters in non-volatile
+  storage devices
+
+## Status
+This document is currently under open discussion. This document is a product of
+[Configuration Task
+Force](https://github.com/RIOT-OS/RIOT/wiki/Configuration-Task-Force-(CTF)), and
+aims to describe the architecture of a runtime configuration system. The content
+of this document is licensed with a Creative Commons CC-BY-SA license.
+
+## Terminology
+This memo uses the [RFC2119](https://www.ietf.org/rfc/rfc2119.txt) terminology
+and the following acronyms and definitions:
+
+### Acronyms
+- RDM: RIOT Developer Memo
+- CTF: RIOT Configuration Task Force
+- RCS: Runtime Configuration System
+- RH: Registry Handler
+- SF: Storage Facility
+
+### Definitions
+- Configuration group: A set of key-value configurations under the same name
+  scheme and scope. E.g `LIGHT_SENSOR_THRESHOLD` and `TRANSMISSION_PERIOD`
+configuration parameters can be contained in an _Application_ configuration
+group, as well as `IEEE802154_CHANNEL` and `IEEE802154_TX_POWER` in a
+_IEEE802.15.4 Radio_ configuration group. Within RIOT, a Configuration Group
+is represented by a Registry Handler.
+
+- Registry Handler: A descriptor that acts as an interface between the RIOT
+Registry and a module that exposes configurations. It provides a common
+interface to get, set, apply and export configurations.
+
+- Storage Facility: A descriptor that acts as an interface between the RIOT
+Registry and a non-volatile storage device. It provides a common interface for
+load/store key-value data from storage devices that might have different
+data representations.
+
+# 1. Introduction
+This document specifies the proposed architecture by the
+Configuration Task Force (CFT) to implement a secure and reliable Runtime
+Configuration System (RCS) focusing on modularity, reuse of existing
+technologies in RIOT (network stack, storage interface) and standard compliance.
+
+# 2. Architecture
+The proposed RCS architecture is formed by a
+[configuration manager](4-configuration-manager) and the
+[RIOT Registry](3-the-riot-registry) and its sub-components:
```

Comment:
would this become more clear if I add "RIOT Registry API" to the box? (instead of just "RIOT Registry")
The RIOT Registry has its own sub-sections RH and SF, and it's stated these are sub-components.

Do you think is necessary to redo the diagrams? Is still not clear?

Reply:
Hmm, then maybe s/RIOT Registry/RIOT Registry Manager? Because it's not just API implementation, it's also stuff that auto-registers RHs/SFs upon boot? And then group all these components in a dotted box labelled "RIOT registry". Yes, I think it would be good to re-do the diagrams accordingly if that's ok

Reply:
Registry Manager seems to be a good name. I will